### PR TITLE
Allow ActiveRecord::QueryMethods#select to accept a hash

### PR DESCRIPTION
**Pull request by alextrueman**
Add ability to use hash with columns and aliases inside #select method.
```ruby
    Post
      .joins(:comments)
      .select(
        posts: { id: :post_id, title: :post_title },
        comments: { id: :comment_id, body: :comment_body}
      )

    # instead of:

    Post
      .joins(:comments)
      .select(
        "posts.id as post_id, posts.title as post_title,
        comments.id as comment_id, comments.body as comment_body"
      )
```
### Summary

Adds processing of `#select` arguments to convert `hash` args into strings to avoid deeper changes of `#select_values`

### Other Information

We have very similar interface for `#where` method, when, for example, we using it with `#joins`:

```ruby
Post.joins(:comments).where(comments: { author_id: 1 })
```
So I think this feature will be pretty obvious and understandable.

### CODE REVIEW

**Comment by simi** [comment]
I did one take on this already at https://github.com/rails/rails/pull/45612, but there was no interest :(.

**Comment by casperisfine** [comment]
> I did one take on this already at #45612, but there was no interest :(.

This is pointing to the current PR.

**Comment by casperisfine** [comment]
I'm gonna assume you meant https://github.com/rails/rails/pull/39874

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 1]
```diff
@@ -315,6 +323,8 @@ def select(*fields)
         return super()
       end
 
+      fields = process_select_args(fields)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:326]: this would make more sense after `check_if_method_has_arguments!`

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 2]
```diff
@@ -1829,6 +1839,34 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
         end
       end
 
+      def process_select_args(fields)
+        fields.map do |field|
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1843]: use `flat_map` over `.map.flatten`

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 3]
```diff
@@ -1829,6 +1839,34 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+          if field.is_a?(Hash)
+            transform_select_hash_values(field)
+          else
+            field
+          end
+        end.flatten
+      end
+
+      def transform_select_hash_values(fields)
+        fields.each_with_object([]) do |(key, columns_aliases), result|
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1853]: This probably should use `flat_map` as well.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 4]
```diff
@@ -1829,6 +1839,34 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+          when Hash
+            columns_aliases.each do |column, column_alias|
+              result << "#{key}.#{column} as #{column_alias}"
+            end
+          when Array
+            columns_aliases.map do |column|
+              result << "#{key}.#{column}"
+            end
+          when String, Symbol
+            result << "#{key} as #{columns_aliases}"
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1864]: I'd need to consult with other members, but I think at least part of these should use Arel rather than generate strings, like in @simi's PR: https://github.com/rails/rails/pull/39874

**Review by casperisfine — COMMENTED** [review]
I think most of this feature make sense and is desirable.

One part I'm a bit on the fence with is `select({ "UPPER(title)" => :title })`, as I fear that accepting raw SQL as hash key may lead to encourage unsafe practice with String interpolation etc. But I may be convinced otherwise. It's also not documented.

And while its scope isn't exactly the same, @simi's implementation in https://github.com/rails/rails/pull/39874 is a bit cleaner in my eyes, notably in its Arel usage. There would probably be value in combining both.

So to summarize: 👍 on the general idea, but this needs a bit of work, and for now I don't feel confident about the SQL fragment hash key part of it (it may be totally fine, but way more thoughts need to be put into it about the security implications).

**Comment by alextrueman** [comment]
Thank you @casperisfine !
I can prepare another PR with Arel implementation ( including your comments ) as well so we can compare both, what do you think?

**Comment by casperisfine** [comment]
Sure. It might also be a good idea to add @simi as co-author on that new PR, as his original one wasn't noticed and got stale for no valid reason.

Also don't hesitate to reduce the scope of your PR, e.g. if you do a PR that only accept symbol arrays and symbol hashes, it will be simple for me to merge it. The SQL fragment feature, I won't feel comfortable making the decision alone. So it might be better to submit it as a follow up PR, and worst case a large part of the feature that cover the most common use cases would already have made it in.

**Comment by alextrueman** [comment]
@casperisfine, 
> Sure. It might also be a good idea to add @simi as co-author on that new PR, as his original one wasn't noticed and got stale for no valid reason.

For sure, this is what I wanted to do 😄 

> Also don't hesitate to reduce the scope of your PR, e.g. if you do a PR that only accept symbol arrays and symbol hashes, it will be simple for me to merge it. The SQL fragment feature, I won't feel comfortable making the decision alone. So it might be better to submit it as a follow up PR, and worst case a large part of the feature that cover the most common use cases would already have made it in.

Funny thing,  but this "SQL fragment feature" was more like a "side-effect" of implementation rather then something I planned to do, I'm on refactoring with Arel so I think this side-effect will gone with this 😅

**Comment by casperisfine** [comment]
> so I can just push it in this, what do you think?

Sure, go ahead.

While we're at it, policy is only one commit per PR unless very special reasons, so you might want to squash/rebase.

**Comment by alextrueman** [comment]
@casperisfine , final question 🙇🏻 

After I refactored with Arel, test for this `select({ "UPPER(title)" => :title })` started failing with error [ on screenshot ]
Should I left it as it is, or it's better to add some validation of field and raise custom message, like: "Use SQL functions with regular SQL strings" ? 👀 

Screenshot 2022-09-01 at 12 30 39

**Comment by casperisfine** [comment]
Well, if we no longer add the feature the test should be removed. 

You screenshot is a bit too truncated for me to really answer your question though. I suggest you remove the test and push what you have once you are happy with it, and then I can look on my side how invalid / unexpected input is handled.

**Comment by alextrueman** [comment]
@casperisfine so I created another PR https://github.com/rails/rails/pull/45921 🙃 Too much commits for rebase 😅

**Comment by simi** [comment]
> @casperisfine so I created another PR #45921 sweat_smile Too much commits for rebase upside_down_face

Why? It was enough to force push in here.

**Comment by alextrueman** [comment]
> > @casperisfine so I created another PR #45921 sweat_smile Too much commits for rebase upside_down_face
> 
> Why? It was enough to force push in here.

Ouch, yeah, I just forget about such option, should I push it here instead and close the new one? 👀

**Comment by simi** [comment]
> > > @casperisfine so I created another PR #45921 sweat_smile Too much commits for rebase upside_down_face
> > 
> > 
> > Why? It was enough to force push in here.
> 
> Ouch, yeah, I just forget about such option, should I push it here instead and close the new one? eyes

I think it would be nice to keep it all together in here with the discussion and code together. :pray:

**Comment by casperisfine** [comment]
> should I push it here instead and close the new one?

It would be best yes, this way we have the full PR discussion in case we git blame these changed lines.

>  started failing with error

Ok, so I looked at it, and I think we must address this. `arel_column` do expect a block, because if it can't map the arguments with an existing column it knows about, it yields back to the caller to provide a SQL fragment.

You can look at how other callers do it and do more or less the same.

This also suggest we should add some test that use a non-existing column so that we cover this branch.

Other than this your patch looks good to me, and I'll merge once this is done.

**Comment by alextrueman** [comment]
@casperisfine , I hope I get your idea correctly 🙇🏻

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 5]
```diff
@@ -1829,6 +1839,39 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+        end
+      end
+
+      def transform_select_hash_values(fields)
+        fields.flat_map do |key, columns_aliases|
+          case columns_aliases
+          when Hash
+            columns_aliases.map do |column, column_alias|
+              arel_column("#{key}.#{column}") do |attr_name|
+                return connection.quote_table_name(attr_name)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1858]:
```suggestion
                connection.quote_table_name(attr_name)
```

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 6]
```diff
@@ -1829,6 +1839,39 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+            end
+          when Array
+            columns_aliases.map do |column|
+              arel_column("#{key}.#{column}") do |attr_name|
+                connection.quote_table_name(attr_name)
+              end
+            end
+          when String, Symbol
+            arel_column(key.to_s) do |attr_name|
+              return connection.quote_table_name(attr_name)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1869]:
```suggestion
              connection.quote_table_name(attr_name)
```

**Review by casperisfine — COMMENTED** [review]
I don't think these returns are correct. You are in the middle of an iterator block that would return the whole method.

File: activerecord/lib/active_record/relation/query_methods.rb (hunk 6 above)
Comment (alextrueman) [reply activerecord/lib/active_record/relation/query_methods.rb:1869]: @casperisfine , without return, here

```ruby
Post.select(foo: :post_title).take
```

instead of
```
ActiveRecord::StatementInvalid: Mysql2::Error: Unknown column 'foo' in 'field list'
```
we will receive:
```
NoMethodError: undefined method `as' for "`foo`":String

            end.as(columns_aliases.to_s)
               ^^^
```

is it okay?

Comment (casperisfine) [reply activerecord/lib/active_record/relation/query_methods.rb:1869]: No that's not ok. You must find a proper way to cast that missing column to proper Arel AST, like `where` does. 

I won't have time to look at this tonight, but if you are stuck just leave a comment and I'll have a look first thing tomorrow.

Comment (alextrueman) [reply activerecord/lib/active_record/relation/query_methods.rb]: @casperisfine, no, not stuck at all, should be fixed now.

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 7]
```diff
@@ -1829,6 +1839,35 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+          end
+        end
+      end
+
+      def transform_select_hash_values(fields)
+        fields.flat_map do |key, columns_aliases|
+          case columns_aliases
+          when Hash
+            columns_aliases.map do |column, column_alias|
+              arel_column("#{key}.#{column}", &:itself).as(column_alias.to_s)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1857]: this `&:itself` is incorrect as it causes `arel_column` to return a String, on which you call `#as`.

I was able to make this fail with this test:

```ruby
    def test_select_with_invalid_nested_field
      assert_raises(ActiveRecord::StatementInvalid) do
        Post.select(posts: { "UPPER(title)" => :post_title }).take
      end
    end
```

File: activerecord/lib/active_record/relation/query_methods.rb [hunk 8]
```diff
@@ -1829,6 +1839,35 @@ def check_if_method_has_arguments!(method_name, args, message = nil)
...
+            end
+          when Array
+            columns_aliases.map do |column|
+              arel_column("#{key}.#{column}", &:itselft)
```
Comment (casperisfine) [inline activerecord/lib/active_record/relation/query_methods.rb:1861]:
```suggestion
              arel_column("#{key}.#{column}", &:itself)
```
typo here, that suggest this path isn't covered by tests.

**Comment by casperisfine** [comment]
```diff
diff --git a/activerecord/lib/active_record/relation/query_methods.rb b/activerecord/lib/active_record/relation/query_methods.rb
index a0d9c3d878f..e1bc3b6e6ee 100644
--- a/activerecord/lib/active_record/relation/query_methods.rb
+++ b/activerecord/lib/active_record/relation/query_methods.rb
@@ -1854,11 +1854,13 @@ def transform_select_hash_values(fields)
           case columns_aliases
           when Hash
             columns_aliases.map do |column, column_alias|
-              arel_column("#{key}.#{column}", &:itself).as(column_alias.to_s)
+              arel_column("#{key}.#{column}") do
+                predicate_builder.resolve_arel_attribute(key.to_s, column)
+              end.as(column_alias.to_s)
             end
           when Array
             columns_aliases.map do |column|
-              arel_column("#{key}.#{column}", &:itselft)
+              arel_column("#{key}.#{column}", &:itself)
             end
           when String, Symbol
             arel_column(key.to_s) do |attribute_name|
diff --git a/activerecord/test/cases/relation/select_test.rb b/activerecord/test/cases/relation/select_test.rb
index 874d4146bb7..0ca53280760 100644
--- a/activerecord/test/cases/relation/select_test.rb
+++ b/activerecord/test/cases/relation/select_test.rb
@@ -33,6 +33,15 @@ def test_select_with_not_exists_field
       end
     end
 
+    def test_select_with_invalid_nested_field
+      assert_raises(ActiveRecord::StatementInvalid) do
+        Post.select(posts: { "UPPER(title)" => :post_title }).take
+      end
+      assert_raises(ActiveRecord::StatementInvalid) do
+        Post.select(posts: ["UPPER(title)"]).take
+      end
+    end
+
     def test_select_with_hash_with_not_exists_field
       assert_raises(ActiveRecord::StatementInvalid) do
         Post.select(posts: { boo: :post_title }).take
```

I fixed this locally, if you apply the above change it's good for me and I'll merge.

**Comment by alextrueman** [comment]
@casperisfine, I'm done, thank you so much for your help on it 🙏🏻

**Comment by casperisfine** [comment]
Welcome. Thanks for the feature.

**Comment by simi** [comment]
Seems my original use-case is not supported. Would be additional PR of supporting raw SQL (`Arel.sql`) accepted?

```ruby
Post.joins(:comments).select('posts.*', Arel.sql('COUNT(comments.id)') => :comment_count).load
/home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `initialize': SQLite3::SQLException: no such column: posts.COUNT(comments.id) (ActiveRecord::StatementInvalid)
        from /home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `new'
        from /home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `prepare'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:48:in `block (2 levels) in exec_query'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:953:in `block in with_raw_connection'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:926:in `with_raw_connection'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:45:in `block in exec_query'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1057:in `log'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:44:in `exec_query'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:589:in `select'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:66:in `select_all'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:110:in `select_all'
        ... 32 levels...
/home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `initialize': no such column: posts.COUNT(comments.id) (SQLite3::SQLException)
        from /home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `new'
        from /home/retro/.gem/ruby/3.1.2/gems/sqlite3-1.4.4/lib/sqlite3/database.rb:152:in `prepare'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:48:in `block (2 levels) in exec_query'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:953:in `block in with_raw_connection'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:926:in `with_raw_connection'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:45:in `block in exec_query'
        from /home/retro/code/work/oss/rails/activesupport/lib/active_support/notifications/instrumenter.rb:58:in `instrument'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:1057:in `log'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/sqlite3/database_statements.rb:44:in `exec_query'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:589:in `select'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:66:in `select_all'
        from /home/retro/code/work/oss/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:110:in `select_all'
        ... 32 levels...

```

**Comment by byroot** [comment]
> Seems my original use-case is not supported.

Yeah, I asked to remove it because:

> One part I'm a bit on the fence with is select({ "UPPER(title)" => :title }), as I fear that accepting raw SQL as hash key may lead to encourage unsafe practice with String interpolation etc.

> Would be additional PR of supporting raw SQL (`Arel.sql`) accepted?

I think having to use `Arel.sql` would cover my reservation about safety, so 👍

**Comment by alextrueman** [comment]
@byroot @simi, I think I have idea how to handle it, do you mind if I will prepare PR for this as well?

**Comment by simi** [comment]
> @byroot @simi, I think I have idea how to handle it, do you mind if I will prepare PR for this as well?

Sure, feel free to.

**Comment by leejarvis** [comment]
Forgive my ignorance, but shouldn't this work for aliases across multiple joins (specifically those that might point to the same table name). For example:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "pg"
end

require "active_record"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "postgresql", database: "select_hash_test")
# ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :authors, force: true do |t|
    t.string :name
  end

  create_table :post_authors, force: true do |t|
    t.belongs_to :post
    t.belongs_to :author
  end

  create_table :comments, force: true do |t|
    t.belongs_to :post
    t.belongs_to :author
    t.string :text
  end
end

class Post < ActiveRecord::Base
  has_many :post_authors
  has_many :authors, through: :post_authors

  has_many :comments
  has_many :comment_authors, through: :comments, source: :author
end

class PostAuthor < ActiveRecord::Base
  belongs_to :post
  belongs_to :author
end

class Author < ActiveRecord::Base
  has_many :post_authors
  has_many :posts, through: :post_authors
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
  belongs_to :author
end

a1 = Author.create!(name: "Author 1")
a2 = Author.create!(name: "Author 2")

post = a1.posts.create!(title: "Post 1")
post.comments.create! text: "Comment 1", author: a2
post.comments.create! text: "Reply", author: a1

p Post
  .select(authors: { name: :author_name }, comment_authors: { name: :commentor_name })
  .joins(:authors, :comment_authors).to_a
```

This code blows up because Rails generates the alias for `comment_authors` as `comment_authors_posts` which of course isn't obvious from the code. It works if I manually set that as the alias: 

```ruby
p Post
  .select(authors: { name: :author_name }, comment_authors_posts: { name: :commentor_name })
  .joins(:authors, :comment_authors).to_a
```
